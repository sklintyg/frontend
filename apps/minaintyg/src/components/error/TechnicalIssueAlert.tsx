import { ErrorIdentifier } from '@frontend/components'
import { IDSAlert } from '@frontend/ids-react-ts'
import { ReactNode } from 'react'
import { QueryError } from '../../utils/isQueryError'

export function TechnicalIssueAlert({
  children,
  headline = 'Tekniskt fel',
  error,
  additionalInfo,
}: {
  children: ReactNode
  headline?: string
  error: QueryError
  additionalInfo?: ReactNode
}) {
  return (
    <IDSAlert headline={headline} type="error">
      <div className="flex flex-col gap-4">
        {children}
        {error.id && (
          <p>
            <ErrorIdentifier id={error.id} centerText={false} />
          </p>
        )}
        {additionalInfo && additionalInfo}
      </div>
    </IDSAlert>
  )
}
