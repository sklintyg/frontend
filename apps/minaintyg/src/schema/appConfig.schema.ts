import { z } from 'zod'

export const appConfigLinkSchema = z.object({
  key: z.string(),
  url: z.string(),
  text: z.string(),
  target: z.string(),
})

export const appConfigSchema = z.object({
  useMinifiedJavascript: z.boolean(),
  version: z.string(),
  buildNumber: z.string(),
  elva77MainUrl: z.string(),
  elva77LoginUrl: z.string(),
  applicationLogoutUrl: z.string(),
  miUserSurveyUrl: z.string(),
  miUserSurveyVersion: z.string(),
  miUserSurveyDateFrom: z.string(),
  miUserSurveyDateTo: z.string(),
  knownRecipients: z.array(z.object({ id: z.string(), name: z.string(), trusted: z.boolean() })),
  links: z.record(appConfigLinkSchema),
  banners: z.array(z.unknown()),
})

export type AppConfig = z.infer<typeof appConfigSchema>
export type AppConfigLink = z.infer<typeof appConfigLinkSchema>
