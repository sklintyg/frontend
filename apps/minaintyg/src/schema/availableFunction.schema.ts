import { z } from 'zod'

export const AvailableFunctionType = z.enum(['SEND_CERTIFICATE'])
export const AvailableFunctionInformationType = z.enum([''])

export const availableFunctionInformationSchema = z.object({
  type: AvailableFunctionInformationType,
  id: z.string(),
})

export const availableFunctionSchema = z.object({
  body: z.string().nullable(),
  description: z.string().nullable(),
  name: z.string().nullable(),
  title: z.string().nullable(),
  type: AvailableFunctionType,
  information: z.array(availableFunctionInformationSchema),
})

export type AvailableFunction = z.infer<typeof availableFunctionSchema>
