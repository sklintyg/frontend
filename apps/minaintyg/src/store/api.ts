import { getCookie } from '@frontend/utils'
import { isAnyOf, isPlainObject } from '@reduxjs/toolkit'
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { AvailableFunction, Certificate, CertificateMetadata } from '../schema/certificate.schema'
import { CertificateFilterOptions } from '../schema/certificateListFilter.schema'
import { User } from '../schema/user.schema'
import { CertificateFilterState } from './slice/certificateFilter.slice'

export const api = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: '/api/',
    prepareHeaders: (headers) => {
      if (getCookie('XSRF-TOKEN')) {
        headers.set('X-XSRF-TOKEN', getCookie('XSRF-TOKEN') ?? '')
      }
      return headers
    },
  }),
  tagTypes: ['User', 'Certificate'],
  endpoints: (builder) => ({
    getUser: builder.query<User, void>({
      query: () => 'user',
      providesTags: ['User'],
    }),
    getCertificates: builder.query<{ content: CertificateMetadata[] }, CertificateFilterState>({
      query: (body) => ({
        url: 'certificate',
        method: 'POST',
        body: Object.fromEntries(Object.entries(body).map(([key, value]) => [key, [value]])),
      }),
      providesTags: (result) =>
        result
          ? [...result.content.map(({ id }) => ({ type: 'Certificate' as const, id })), { type: 'Certificate', id: 'LIST' }]
          : [{ type: 'Certificate', id: 'LIST' }],
    }),
    getCertificatesFilter: builder.query<CertificateFilterOptions, void>({
      query: () => 'filters',
    }),
    getCertificate: builder.query<Certificate & { availableFunctions: AvailableFunction[] }, { id: string }>({
      query: ({ id }) => `certificate/${id}`,
      transformResponse: ({ certificate, availableFunctions }: { certificate: Certificate; availableFunctions: AvailableFunction[] }) => ({
        ...certificate,
        availableFunctions,
      }),
      providesTags: (result) => (result ? [{ type: 'Certificate' as const, id: result.metadata.id }] : []),
    }),
    sendCertificate: builder.mutation<void, { id: string }>({
      query: ({ id }) => ({
        url: `certificate/${id}/send`,
        method: 'POST',
      }),
      invalidatesTags: (result, error, { id }) => (error ? [] : [{ type: 'Certificate', id }]),
    }),
  }),
})

export const { useGetCertificatesQuery, useGetCertificateQuery, useGetCertificatesFilterQuery, useSendCertificateMutation } = api

export const isFulfilledEndpoint = isAnyOf(...Object.values(api.endpoints).map((endpoint) => endpoint.matchFulfilled))
export const isRejectedEndpoint = isAnyOf(...Object.values(api.endpoints).map((endpoint) => endpoint.matchRejected))
export const hasResponse = (o: unknown): o is { response: Response } =>
  isPlainObject(o) && 'response' in o && o.response instanceof Response
