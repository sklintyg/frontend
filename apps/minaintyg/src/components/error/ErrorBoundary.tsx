import { randomUUID } from '@frontend/utils'
import { useMemo } from 'react'
import { ErrorPageHero } from './ErrorPageHero'

export function ErrorBoundary() {
  const id = useMemo(randomUUID, [])
  // TODO: useEffect hook that logs error back to the backend.
  return <ErrorPageHero id={id} type="unknown" />
}
