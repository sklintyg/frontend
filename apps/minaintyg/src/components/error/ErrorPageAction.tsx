import { getNavigationItem, getNavigationItemUrl } from '@frontend/components/1177'
import { IDSIconChevron, IDSIconExternal, IDSLink } from '@frontend/ids-react-ts'
import { Link } from 'react-router-dom'
import type { ErrorPageActionTypeEnum } from '../../schema/error.schema';
import { ErrorPageActionType } from '../../schema/error.schema'

export function ErrorPageAction({ type }: { type: ErrorPageActionTypeEnum }) {
  const startLinkItem = getNavigationItem('Start')

  if (type === ErrorPageActionType.enum.start) {
    return (
      startLinkItem && (
        <IDSLink key="start">
          <IDSIconChevron />
          <Link to={getNavigationItemUrl(startLinkItem, import.meta.env.MODE)}>Till startsidan</Link>
        </IDSLink>
      )
    )
  }

  if (type === ErrorPageActionType.enum.login) {
    return (
      <IDSLink key="login">
        <IDSIconChevron />
        <a href="/">Till inloggning</a>
      </IDSLink>
    )
  }

  if (type === ErrorPageActionType.enum[1177]) {
    return (
      <IDSLink key="inera">
        <IDSIconChevron />
        <Link to="https://www.1177.se">Till 1177</Link>
        <IDSIconExternal slot="append-icon" />
      </IDSLink>
    )
  }

  if (type === ErrorPageActionType.enum.about) {
    return (
      <IDSLink key="inera">
        <IDSIconChevron />
        <Link to="https://www.1177.se/om-1177/nar-du-loggar-in-pa-1177.se/det-har-kan-du-gora-nar-du-loggat-in/hantera-intyg/hantera-dina-lakarintyg-och-lakarutlatanden/">
          Om 1177 intyg
        </Link>
        <IDSIconExternal slot="append-icon" />
      </IDSLink>
    )
  }
}
