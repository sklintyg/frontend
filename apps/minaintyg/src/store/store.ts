import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
