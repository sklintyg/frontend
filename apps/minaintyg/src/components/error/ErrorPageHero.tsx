import { ErrorIdentifier } from '@frontend/components'
import { PageHero, PageHeroActions } from '@frontend/components/1177'
import { ErrorPageActionType, ErrorType, ErrorTypeEnum } from '../../schema/error.schema'
import { ErrorPageAction } from './ErrorPageAction'
import { SupportLink } from './SupportLink/SupportLink'

const errorInfo = {
  [ErrorType.enum.unavailable]: {
    heading: 'Tjänsten är inte tillgänglig just nu',
    description: 'Välkommen tillbaka vid senare tillfälle.',
    actions: [],
  },
  [ErrorType.enum['login-failed']]: {
    heading: 'Inloggning misslyckades',
    description: 'På grund av ett tekniskt fel går det inte att logga in just nu. Försök igen senare.',
    actions: [ErrorPageActionType.enum.start],
  },
  [ErrorType.enum['logged-out']]: {
    heading: 'Du är utloggad',
    description:
      'Du har antingen valt att logga ut eller blivit utloggad på grund av inaktivitet. Du kan nu stänga fliken eller logga in på nytt.',
    actions: [ErrorPageActionType.enum.login, ErrorPageActionType.enum[1177]],
  },
  [ErrorType.enum['not-found']]: {
    heading: 'Sidan hittades inte',
    description: 'Sidan du söker har fått en ny adress eller är borttagen.',
    actions: [ErrorPageActionType.enum.start],
  },
  [ErrorType.enum.underage]: {
    heading: 'Du som är under 16 har inte tillgång till 1177 intyg',
    description: 'Vänd dig till din mottagning om du har frågor eller behöver en kopia av ditt intyg.',
    actions: [ErrorPageActionType.enum.start, ErrorPageActionType.enum.about],
  },
  [ErrorType.enum.unknown]: {
    heading: 'Någonting gick fel',
    description: 'På grund av ett tekniskt fel går det inte att visa Intyg just nu. Försök igen senare.',
    actions: [ErrorPageActionType.enum.start],
  },
} as const

function getErrorHeading(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.heading) || errorInfo.unknown.heading
}

function getErrorDescription(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.description) || errorInfo.unknown.description
}

function getActions(type?: ErrorTypeEnum) {
  return (type !== undefined && errorInfo[type]?.actions) || errorInfo.unknown.actions
}

export function ErrorPageHero({ id, type }: { id?: string; type?: ErrorTypeEnum }) {
  return (
    <PageHero heading={getErrorHeading(type)} type={type === ErrorType.enum['logged-out'] ? 'success' : 'error'}>
      <p className="ids-preamble">{getErrorDescription(type)}</p>
      <PageHeroActions>
        {getActions(type).map((t) => (
          <ErrorPageAction type={t} key={t} />
        ))}
      </PageHeroActions>
      {id && type !== ErrorType.enum['logged-out'] && (
        <div className="border-y border-stone-clear py-5">
          Om problemet kvarstår, spara nedan id och kontakta <SupportLink />
          <br />
          <ErrorIdentifier id={id} showTitle={false} />
        </div>
      )}
    </PageHero>
  )
}
