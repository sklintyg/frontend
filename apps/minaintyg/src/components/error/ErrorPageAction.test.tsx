import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import { ErrorPageActionType } from '../../schema/error.schema'
import { api } from '../../store/api'
import { startSession } from '../../store/slice/session.slice'
import { store } from '../../store/store'
import { ErrorPageAction } from './ErrorPageAction'

it.each(ErrorPageActionType.options)('Should render %s type action', async (type) => {
  store.dispatch(startSession())
  await store.dispatch(api.endpoints.getInfo.initiate())
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <ErrorPageAction type={type} />
      </MemoryRouter>
    </Provider>
  )
  expect(container).toMatchSnapshot()
})
