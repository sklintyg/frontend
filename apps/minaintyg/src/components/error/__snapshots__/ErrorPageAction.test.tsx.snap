// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render 1177 type action 1`] = `
<div>
  <a
    class="ids-link ids-link--no-underline"
    href="https://www.1177.se"
    rel="noreferrer"
    target="_blank"
  >
    <span
      class="ids-icon-arrow-right-small  ids-icon--text-start"
    />
    Till 1177
    <span
      class="ids-icon-external-link-small  ids-icon--text-end"
    />
  </a>
</div>
`;

exports[`Should render about type action 1`] = `
<div>
  <a
    class="ids-link ids-link--no-underline"
    href="https://www.1177.se/om-1177/nar-du-loggar-in-pa-1177.se/det-har-kan-du-gora-nar-du-loggat-in/hantera-intyg/hantera-dina-lakarintyg-och-lakarutlatanden/"
    rel="noreferrer"
    target="_blank"
  >
    <span
      class="ids-icon-arrow-right-small  ids-icon--text-start"
    />
    Om 1177 intyg
    <span
      class="ids-icon-external-link-small  ids-icon--text-end"
    />
  </a>
</div>
`;

exports[`Should render login type action 1`] = `
<div>
  <a
    class="ids-link ids-link--no-underline"
    href="/"
    target="self"
  >
    <span
      class="ids-icon-arrow-right-small  ids-icon--text-start"
    />
    Till inloggning
  </a>
</div>
`;

exports[`Should render start type action 1`] = `
<div>
  <a
    class="ids-link ids-link--no-underline"
    href="https://e-tjanster.1177.se/mvk/"
  >
    <span
      class="ids-icon-arrow-right-small  ids-icon--text-start"
    />
    Till startsidan
  </a>
</div>
`;
